eva/optimizer/plan_generator.py

```diff
@@ -131,9 +133,12 @@ def _recursive_strip_exchange(plan: AbstractPlan, is_top: bool = False):
         else:
             return physical_plan
 
-    def build(self, logical_plan: Operator):
+    async def build(self, logical_plan: Operator):
         # apply optimizations
-        plan = self.optimize(logical_plan)
+        try:
+            plan = await wait_for(self.optimize(logical_plan), timeout=60.0)
```

Comment:
Do we decide we want to use `asyncio` to achieve this goal? If so, I think we need to wrap the complete processing pipeline in async, instead of calling `asyncio.run`. Or we can use `loop.run_until_complete`. BTW, how do we ensure the plan we get is valid when it times out?

Reply:
We throw an error if we time out, so EVA crashes.
There are two alternatives to `asyncio`:
- Using `signal`, but this only has UNIX support
- Using a third party software, which provides timeout support using decoraters, but it needs to be installed to be used

Reply:
How would `loop.run_until_complete` be different from `asyncio.run`? Don't they both use event loops in the end?

Reply:
In Postgres, a basic optimization algorithm is used to come up with a valid plan. After that, the optimizer keeps iteratively refining it using genetic programming -- and this process can be stopped at any time to get a valid plan. We will need a boolean variable to check if a valid plan is constructed (first stage is completed).

Reply:
try:
    loop.run_until_complete(asyncio.wait_for(my_coroutine(), timeout=3))
except asyncio.TimeoutError:
    print("Coroutine took too long to complete!")
finally:
    loop.close()

Reply:
asyncio.run creates a new event loop by default, while loop.run_until_complete uses the current event loop

Reply:
I feel this is non-trivial feature and may require some re-design of the optimizer architecture. Is this priority? I think we do not have complicated rules now that will cause the optimizer to run over 60 seconds. @gaurav274 What do you think?

Reply:
My motivation behind this feature is to throw an error if we cannot construct a valid plan for the query. Ideally, this should not occur but due to some bug in the optimizer or poor handling in the binder, we might reach such a state. Can we accomplish this use case without any changes to the optimizer?

Reply:
@xzdandy signal has only UNIX support, which is a problem.

Reply:
You are right. Let's go for asyncio approach.